[Codegen, X86] Add prefetch insertion based on Propeller profile

rlavaee opened the pull request:
This PR implements the prefetch insertion in the InsertCodePrefetch pass based on the [RFC](https://discourse.llvm.org/t/rfc-code-prefetch-insertion/88668).

If the prefetch target is not defined in the same module (i.e, prefetch target function is not defined in the same module), we emit a fallback weak symbol after the prefetch instruction so that if the symbol is not ever defined, we don't get undefined symbol error and the prefetch instruction prefetches the next address:

```
prefetchit1 __llvm_prefetch_target_foo(%rip)
.weak __llvm_prefetch_target_foo
__llvm_prefetch_target_foo:
```

The weak symbol semantic is tied to ELF, so this makes this PR target-dependent.

RKSimon commented on llvm/lib/Target/X86/PrefetchInsertion.cpp:
```diff
@@ -0,0 +1,209 @@
+//===-- PrefetchInsertion.cpp ---=========-----------------------------===//
```
(style) X86PrefetchInsertion.cpp

RKSimon commented on llvm/include/llvm/CodeGen/BasicBlockSectionsProfileReader.h:
```diff
@@ -42,6 +42,17 @@ struct BBClusterInfo {
   unsigned PositionInCluster;
 };
 
+struct BBPosition {
```
(style) public data structures need dxoygen descriptions

RKSimon commented on llvm/include/llvm/CodeGen/MachineBasicBlock.h:
```diff
@@ -348,24 +356,24 @@ class MachineBasicBlock
-  reverse_instr_iterator       instr_rend  ()       { return Insts.rend();   }
-  const_reverse_instr_iterator instr_rend  () const { return Insts.rend();   }
+  reverse_instr_iterator instr_rend() { return Insts.rend(); }
+  const_reverse_instr_iterator instr_rend() const { return Insts.rend(); }
```
(style) don't clang-format lines in a file unaffected by a patch

RKSimon reviewed (changes requested):
blanket clang-format makes it difficult to see the bones of this patch - and there are very few useful comments and descriptions on what is going on

boomanaiden154 commented on llvm/lib/Target/X86/CMakeLists.txt:
```diff
@@ -1,126 +1,113 @@
 add_llvm_component_group(X86 HAS_JIT)
 
-set(LLVM_TARGET_DEFINITIONS X86.td)
+    set(LLVM_TARGET_DEFINITIONS X86.td)
```
It looks like you tried to autoformat this? The changes here fail the build on premerge.

rlavaee committed:
feat(AsmPrinter): Add support for inserting code prefetch instructions.

boomanaiden154 reviewed (commented):
Can we also get a NewPM version of this pass? See some of my recent patches to https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86PassRegistry.def on how to wire it up.

rlavaee committed:
Update tests.

rlavaee commented:
> https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86PassRegistry.def

The Prefetch insertion pass will be architecture-independent.

rlavaee committed:
clang-format.

rlavaee committed:
Remove unwanted changes.

rlavaee committed:
[CodeGen] Centralize prefetch target storage in MachineFunction

- Move prefetch target callsite indexes from MachineBasicBlock to a
  centralized DenseMap in MachineFunction, renamed to PrefetchTargets.
- Change PrefetchTargets type to DenseMap<UniqueBBID, SmallVector<unsigned>>
  to support both mapped and dangling targets.
- Refactor AsmPrinter to use member functions emitPrefetchTargetSymbol
  and emitDanglingPrefetchTargets for consistent emission.
- Update MIR serialization to use the new PrefetchTargets map and
  rename the YAML field to 'prefetch-targets'.
- Update the MIR format for bb_id to 'bb_id <base-id>, <clone-id>, <callsite-index>'
  for improved readability and to avoid ambiguity with floating point.
- Update related tests and add MIR parsing error test cases.

rlavaee committed:
clang-format.

rlavaee committed:
Additional changes.

rlavaee committed:
Split setPrefetchTargets and insertPrefetchHints into separate functions.

rlavaee committed:
clang-format.

rlavaee committed:
Implement fallback support.

rlavaee committed:
clang-format.

rlavaee committed:
fix the prefetch test.

rlavaee committed:
Change undef to extfunc.

rlavaee committed:
Add include for MCContext.h

rlavaee commented:
This PR is now ready for review. I appreciate your feedback.

rlavaee committed:
nits.

boomanaiden154 commented on llvm/include/llvm/CodeGen/InsertCodePrefetch.h:
```diff
@@ -0,0 +1,25 @@
+//===- BasicBlockSectionUtils.h - Utilities for basic block sections     --===//
```
Filename doesn't match the header here.

boomanaiden154 commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -22,6 +22,7 @@
 #include "llvm/ADT/SmallPtrSet.h"
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/Statistic.h"
+#include "llvm/ADT/StringExtras.h"
```
This looks unnecessary for this patch?

rlavaee committed:
Address comments.

rlavaee committed:
clang-format.

rlavaee commented on llvm/lib/CodeGen/BasicBlockSectionsProfileReader.cpp:
```diff
@@ -101,6 +101,14 @@ BasicBlockSectionsProfileReader::getPrefetchTargetsForFunction(
+SmallVector<PrefetchHint>
+BasicBlockSectionsProfileReader::getPrefetchHintsForFunction(
+    StringRef FuncName) const {
+  auto R = ProgramOptimizationProfile.find(getAliasName(FuncName));
```
Done.

tmsri commented on llvm/lib/CodeGen/BasicBlockSectionsProfileReader.cpp:
```diff
@@ -367,6 +395,45 @@ Error BasicBlockSectionsProfileReader::ReadV1Profile() {
+      Values[1].split(PrefetchTargetStr, ',');
+      if (PrefetchTargetStr.size() != 3)
+        return createProfileParseError(
+            Twine("Prefetch target target expected: ") + Values[1]);
```
Used target twice.

rlavaee committed:
Address comments.

rlavaee committed:
clang-format.

rlavaee commented on llvm/lib/CodeGen/BasicBlockSectionsProfileReader.cpp:
```diff
@@ -186,6 +203,17 @@ BasicBlockSectionsProfileReader::getPrefetchTargetsForFunction(
 // +----------------------------------+
 //
+// A prefetch hint specified in function "bar" as "120,1 foo,10,2" results
+// in a hint inserted after the first call in block #120 of bar:
```
Done.

rlavaee committed:
Fix post-instruction symbol handling

The order of the if statements had been reversed which was incorrect.

rlavaee committed:
Use llvm:: for the getPrefetchTargetSymbolName function definition

rlavaee committed:
Add new header style and set runOnMachineFunction's return value.

rlavaee committed:
Fix the defined function target check.

HaohaiWen reviewed (approved):
LGTM

boomanaiden154 reviewed (approved):
LGTM.

It would be nice to add a NewPM port of `InsertCodePrefetch`, but given that pass already exists, we can just do that in a future PR when it comes up.

MaskRay commented on llvm/include/llvm/CodeGen/InsertCodePrefetch.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/Support/CommandLine.h"
```
unused STLExtras.s and CommandLine.h

rlavaee replied:
Done.

MaskRay reviewed (approved):
MC usage looks good.

rlavaee committed:
Fix the includes.

rlavaee commented:
> LGTM.
> 
> It would be nice to add a NewPM port of `InsertCodePrefetch`, but given that pass already exists, we can just do that in a future PR when it comes up.

Is this still relevant given that the pass is now arch-independent (defined in Codegen/)?

rlavaee committed:
Fix the comment.

rlavaee committed:
Merge branch 'main' into prefetch-insertion

ro-i commented:
@rlavaee This PR seems to have introduced a sometimes-occurring test failure in https://lab.llvm.org/buildbot/#/builders/187.
`LLVM::basic-block-sections-code-prefetch.ll`
